plugin/path_checkouts.go

```diff
@@ -5,11 +5,325 @@ import (
+	"github.com/hashicorp/go-multierror"
 	"github.com/hashicorp/vault/sdk/framework"
 	"github.com/hashicorp/vault/sdk/logical"
 )
 
-func (b *backend) pathReserveStatus() *framework.Path {
+const secretAccessKeyType = "creds"
+
+// TODO we're going to need to deal with mutexes.
+func (b *backend) pathSetCheckOut() *framework.Path {
+	return &framework.Path{
+		Pattern: libraryPrefix + framework.GenericNameRegex("name") + "/check-out$",
+		Fields: map[string]*framework.FieldSchema{
+			"name": {
+				Type:        framework.TypeLowerCaseString,
+				Description: "Name of the set",
+				Required:    true,
+			},
+			"ttl": {
+				Type:        framework.TypeDurationSecond,
+				Description: "In seconds, the length of time before the check-out will expire.",
+			},
+		},
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.ReadOperation: &framework.PathOperation{
+				Callback: b.operationSetCheckOut,
+				Summary:  "Check a service account out from the library.",
+			},
+		},
+		HelpSynopsis: `Check a service account out from the library.`,
+	}
+}
+
+func (b *backend) operationSetCheckOut(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	setName := fieldData.Get("name").(string)
+
+	ttlPeriodRaw, ttlPeriodSent := fieldData.GetOk("ttl")
+	if !ttlPeriodSent {
+		ttlPeriodRaw = 0
+	}
+	requestedTTL := time.Duration(ttlPeriodRaw.(int)) * time.Second
+
+	set, err := readSet(ctx, req.Storage, setName)
+	if err != nil {
+		return nil, err
+	}
+	if set == nil {
+		return logical.ErrorResponse(`"%s" doesn't exist`, setName), nil
+	}
+
+	// Find the first service account available.
+	var availableServiceAccountName string
+	for _, serviceAccountName := range set.ServiceAccountNames {
+		checkOut, err := b.checkOutHandler.Status(ctx, req.Storage, serviceAccountName)
+		if err != nil {
+			return nil, err
+		}
+		if checkOut.IsAvailable {
+			availableServiceAccountName = serviceAccountName
+			break
+		}
+	}
+	if availableServiceAccountName == "" {
+		resp, err := logical.RespondWithStatusCode(&logical.Response{
+			Warnings: []string{"No service accounts available for check-out, please try again later."},
+		}, req, 429)
+		if err != nil {
+			return nil, err
+		}
+		return resp, nil
+	}
+	ttl := set.TTL
+	if ttlPeriodSent {
+		switch {
+		case set.TTL <= 0 && requestedTTL > 0:
+			// The set's TTL is infinite and the caller requested a finite TTL.
+			ttl = requestedTTL
+		case set.TTL > 0 && requestedTTL < set.TTL:
+			// The set's TTL isn't infinite and the caller requested a shorter TTL.
+			ttl = requestedTTL
+		}
+	}
+	due := time.Now().UTC().Add(ttl)
+	if ttl <= 0 {
+		// An unlimited lending period is allowed. For simplicity, set
+		// the due date to be 100 years in the future.
+		// The due date is mainly for use during "status" calls so it's
+		// easy to understand who has checked something out and for how long.
+		due = time.Now().UTC().Add(time.Hour * 24 * 365 * 100)
```

Comment:
Is `due` strictly for display purposes (e.g. no enforcement, like TTL)?
Will this result in displaying "Due in 99.98 years" or something like that? If so, I think it would be cleaner to store a sentinel time here (0 would work), and then report out expiration as `∞` like we do for non-expiring tokens.

Reply:
Actually, upon reflecting on it more, I think we have no choice but to strip out `due` even though it's theoretically nice for UX. If an infinite time is chosen here, it'll be overridden by the leasing framework to be `32d` or whatever the outer settings are. I don't see a clean way to learn what the outer settings are from within the context of a request. It might be possible with a lease ID, but the lease ID is only available _after_ the response passes through higher levels. Or, maybe with a `SystemView` to look up TTLs and then try to duplicate the logic for the lease durations, but that's also not great.

Reply:
I've pushed a commit stripping out the "due" field. Instead folks will need to rely upon lease durations.